PLT-1222 ADR about changing the slot to time conversions in the emulator and PAB (#829)
input-output-hk/plutus-apps: doc/adr/0015-time-conversion.rst

through to the ``Contract`` API, which users can use to convert between a ``Slot`` and a ``POSIXTime``.
However, the current ``SlotConfig`` representation supposes that the slot length is the same for all
epochs in the Cardano blockchain, which is not the case.
For example, during the Byron era, the slot length was 20s, while from Shelley era and onwards, the
slot length is 1s.
Therefore, the functions from the ``Ledger.TimeSlot`` module in ``plutus-ledger`` do not compute the
conversion between ``Slot`` and ``POSIXTime`` the right way.
The current easiest way to compute the time conversions is to query the local Cardano node on the
consensus layer, which requires the ``ouroboros-consensus`` dependency.

Decision
--------

* We will deprecate the ``Ledger.TimeSlot.SlotConfig`` type and all functions in the
  ``Ledger.TimeSlot`` module using the ``SlotConfig``. The only viable functions are the ones that
  convert between ``Data.Time`` types and plutus types (types related to ``TxInfo``).

* We will copy the ``Ledger.TimeSlot`` module in the emulator (ideally rename it) and keep it as
  an internal module. Any functions not used by the emulator will be removed.

> * We will move the ``Ledger.Params`` module inside the emulator as an internal module and modify the

Reviewer: Maybe we can just remove `SlotConfig` from `Params`. It doesn't need to be configurable anymore.

Reviewer 2: Developpers may want to test how there contract behave with a given `SlotConfig`?

Author: I think it *does* need to be configurable. The PAB simulator actually uses it to define the `delay` for creating new blocks.

Reviewer: Ok!

  ``Params`` datatype name to ``EmulatorParams``.

* We will modify the ``Plutus.Contract.Request.getParams`` function to

>   ``Plutus.Contract.Request.getProtocolParameters``. This implies to modify the name of Contract

Reviewer 3: ```suggestion
  ``Plutus.Contract.Request.getProtocolParameters``. This implies modifying the name of Contract
```

  effect ``GetParamsReq/GetParamsResp``.

* We will create two pairs of effects in ``Plutus.Contract.Effect``:

  .. code-block:: haskell

    data PABReq =
      ...
      | SlotToUTCTimeIntervalReq SlotNo
      | UTCTimeToSlotReq UTCTime
      ...

     data PABResp =
      ...
      | SlotToUTCTimeIntervalResp (UTCTime, SlotLength) -- An alternative can be (UTCTime, UTCTime)
      | UTCTimeToSlotResp SlotNo
      ...

* We will implement the emulator effect interpreter by simply using the ``SlotConfig`` for the conversions.

* We will implement the PAB effect interpreter by using the local node. There are multiple steps to
  implement this:

  * At startup, the PAB will query the ``EraHistory`` from the local node and store it in its local
    environment.

>   * We will implement the PAB interpreter by using the ``EraHistory`` alonside the consensus

Reviewer 3: ```suggestion
  * We will implement the PAB interpreter by using the ``EraHistory`` alongside the consensus
```

        (relativeTime, _) <- interpretQuery interpreter $ slotToWallclock slotNo
        pure $ fromRelativeTime systemStart relativeTime

    However, we will also add an additional step. If the conversion returns ``PastHorizonException``,
    then there is a good probability that the ``EraHistory`` is out of date.
    The reason is that ``EraHistory`` only encodes era information from the moment the user ran the
    query and it cannot predict the future.
    In that case, if the ``PastHorizonException`` is returned, we will re-query the ``EraHistory``
    of the local node, replace the old value in the PAB environment, and retry the conversion.
    If it fails again, we return the error message.

Argument
--------

The solution to create new effects in ``Plutus.Contract.Effects`` has the nice property that the
implementation can differ depending on the environment.
It allows us to keep using ``SlotConfig`` in the emulator, while using the existing implementation
provided by ``ouroboros-consensus`` when using a real Cardano node.

Implications
------------

* We will have to update the ``plutus-use-cases`` examples to use those new conversion functions.

>   The user will **not** use ``SlotConfig`` to work convert between slots and UTC time. He will

Reviewer 3: ```suggestion
  The user will **not** use ``SlotConfig`` to convert between slots and UTC time. He will
```
